Source/Details/ASPageTable.h

```diff
@@ -0,0 +1,82 @@
+//
+//  ASPageTable.h
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import <AsyncDisplayKit/ASBaseDefines.h>
+
+@class ASCollectionElement;
+
+NS_ASSUME_NONNULL_BEGIN
+
+AS_SUBCLASSING_RESTRICTED
+@interface ASPageCoordinates : NSObject
+
+@property (nonatomic, assign, readonly) NSUInteger x;
+@property (nonatomic, assign, readonly) NSUInteger y;
+
++ (instancetype)pageCoordinatesWithX:(NSUInteger)x Y:(NSUInteger)y;
```

Comment:
Since these are going to be created _a lot_ I think it's worth optimizing upfront. Let's pack the coordinates into an integer and trick NSMapTable into using it.

```objc
typedef struct {
  uint16_t x,
  uint16_t y
} ASPageCoordinate;

__unsafe_unretained id ASPageCoordinateMakeKey(ASPageCoordinate coordinate) {
  return (__bridge id)(void *)((coordinate.x << 16) + coordinate.y + 1);
}
```

- We add 1 because 0 is not a valid key for NSMapTable.
- It might be worth asserting that `x, y < (1 << 16)` and masking `x,y & 0xFFFF`.

Then create the map table with `pointerFunctionsWithOptions:NSPointerFunctionsIntegerPersonality | NSPointerFunctionsOpaqueMemory` and use `[mapTable objectForKey:ASPageCoordinateMakeKey(coord)]` to get your value.